jcl/src/java.base/share/classes/jdk/internal/misc/Unsafe.java

```diff
@@ -4092,6 +4324,68 @@ public final Object getAndSetObjectAcquire(Object obj, long offset, Object value
 		}
 	}
 
+/*[IF Java12]*/
+	/**
+	 * Atomically sets value at offset in obj
+	 * and returns the value of the field prior to the update.
+	 * The get operation has the memory semantics of getVolatile.
+	 * The set operation has the memory semantics of setVolatile.
+	 * 
+	 * @param obj object into which to set the value
+	 * @param offset location to set value in obj
+	 * @param value to set in obj memory
+	 * @return value of field in obj at offset before update
+	 */
+	public final Object getAndSetReference(Object obj, long offset, Object value) {
+		for (;;) {
+			Object objectAtOffset = getReferenceVolatile(obj, offset);
+			if (weakCompareAndSetReference(obj, offset, objectAtOffset, value)) {
```

Comment:
`weakCompareAndSetReference` is this spec'd to have `weak` memory ordering effects?  Then it won't meet the `get/setVolatile` semantics.

Reply:
`weakCompareAndSetReference` just calls the `compareAndSetReference` native. I can update this method with `compareAndSetReference` to make the memory ordering more clear.

Reply:
We should call the method defined to have the memory semantics we want as the JIT may optimize some of these calls.  It would introduce a bug if we call the "wrong" one because we know the implementation and the JIT replaces it with an intrinsics with different memory semantics

Reply:
ah okay. in that case the other `getAndSetXXX` methods are wrong as well. I will fix them in a separate pr

Reply:
Why a separate PR?

Reply:
mostly because this pr has only java 12 changes. i'll change it here if you prefer

Reply:
My understanding was that @DanHeidinga was pointing out changes required to methods _added_ by this PR for Java 12. I'd prefer that we don't start with versions we know are wrong.

Reply:
Fix the Java 12 ones in this PR.  The other APIs can be a separate one.

I'd like to see our 12 builds up and running asap